onnx/defs/nn/defs.cc

```diff
@@ -2829,4 +2829,290 @@ ONNX_OPERATOR_SET_SCHEMA(
+                                                                                                  // sequence_length,
+                                                                                                  // rotary_embedding_dim
+                                                                                                  // / 2]
+              .Add("CosCacheUnsqueezed = Unsqueeze(CosCacheSliced, Two1D)") // shape of cos matrix: [batch_size,
+                                                                            // sequence_length, 1, rotary_embedding_dim
+                                                                            // / 2]
+              .Add("SinCacheUnsqueezed = Unsqueeze(SinCacheSliced, Two1D)"); // shape of sin matrix: [batch_size,
+                                                                             // sequence_length, 1, rotary_embedding_dim
+                                                                             // / 2]
+
+          // Create slices of inputs to multiply with sin and cos matrices based on interleaved parameter
+          // For non-interleaved (basic) rotation, slices are created as follows,
+          builder.Add(
+              "X1Basic, X2Basic = Split <axis = -1, num_outputs = 2> (XToRotate)"); // shape of X1 =
+                                                                                    // input[:,:,:,:rotary_embedding_dim/2],
+                                                                                    // X2 =
+                                                                                    // input[:,:,:,rotary_embedding_dim/2:rotary_embedding_dim]
+          // For interleaved rotation, slices are created as follows,
+          builder.Const1D("One1D", (int64_t)1)
+              .Const1D("AxesRotaryDim", (int64_t)3)
+              .Add("RotaryEmbedDimInclusive = Add(RotaryEmbedDim, One1D)")
+              .Add(
+                  "X1Interleaved = Slice(XToRotate, Zero1D, RotaryEmbedDim, AxesRotaryDim, Two1D)") // shape of X1 =
+                                                                                                    // input[:,:,:,0:rotary_embedding_dim:2]
+              .Add(
+                  "X2Interleaved = Slice(XToRotate, One1D, RotaryEmbedDimInclusive, AxesRotaryDim, Two1D)"); // shape of
+                                                                                                             // X2 =
+                                                                                                             // input[:,:,:,1:rotary_embedding_dim:2]
+
+          // Choose the correct slices based on interleaved parameter
+          // real = cos_x * x1 - sin_x * x2
+          // imag = sin_x * x1 + cos_x * x2
+          builder.Const1D("InterleavedParam", interleaved)
+              .Add("InterleaveCond = Equal(InterleavedParam, Zero1D)")
```

Comment:
Hmmm, I added some comments here previously ... not sure what happened to them!

Since we know the interleaved attribute value (when this code executes), we should just use if-then-else to handle the interleaved and non-interleaved case separately (constructing different function bodies in the two cases). Otherwise, we end up with a more complex function body that will be less-performant (unless another optimizer manages to eliminate it).

Reply:
Modified.